Marlin/Marlin_main.cpp

```diff
@@ -1429,6 +1434,108 @@ static void setup_for_endstop_move() {
+       sync_plan_position();
+       gcode_M241(1000);
+       bedlevelprobes[0] = gcode_M238(4);
+     
+       do_blocking_move_to(current_position[X_AXIS], current_position[Y_AXIS] + 0.1, z);
+       sync_plan_position();
+       gcode_M241(350);
+       bedlevelprobes[1] = gcode_M238(4);
+     
+       do_blocking_move_to(current_position[X_AXIS] + 0.1, current_position[Y_AXIS], z);
+       sync_plan_position();
+       gcode_M241(350);
+       bedlevelprobes[2] = gcode_M238(4);
+     
+       do_blocking_move_to(current_position[X_AXIS] + 0.1, current_position[Y_AXIS], z);
+       sync_plan_position();
+       gcode_M241(350);
+       bedlevelprobes[3] = gcode_M238(4);
+     
+       do_blocking_move_to(current_position[X_AXIS], current_position[Y_AXIS] - 0.1, z);
+       sync_plan_position();
+       gcode_M241(350);
+       bedlevelprobes[4] = gcode_M238(4);
+     
+       do_blocking_move_to(current_position[X_AXIS], current_position[Y_AXIS] - 0.1, z);
+       sync_plan_position();
+       gcode_M241(350);
+       bedlevelprobes[5] = gcode_M238(4);
+     
+       do_blocking_move_to(current_position[X_AXIS] - 0.1, current_position[Y_AXIS], z);
+       sync_plan_position();
+       gcode_M241(350);
+       bedlevelprobes[6] = gcode_M238(4);
+     
+       do_blocking_move_to(current_position[X_AXIS] - 0.1, current_position[Y_AXIS], z);
+       sync_plan_position();
+       gcode_M241(350);
+       bedlevelprobes[7] = gcode_M238(4);
+     
+       do_blocking_move_to(x, y, z);
+       sync_plan_position();
+       gcode_M241(350);
+       bedlevelprobes[8] = gcode_M238(4);
+     
+       uint16_t num_samples = 0x0001 << 3;
+       uint16_t i = 0;
+       uint32_t sample_sum = 0;
+       uint16_t sample_avg = 0;
+     
+       for(i = 0; i < 8; i++) {
+         sample_sum += bedlevelprobes[i];
+         if (verbose_level > 3) {
+           SERIAL_PROTOCOLPGM("sample_point #");
+           SERIAL_PROTOCOL_F(i, 10);
+           SERIAL_PROTOCOLPGM(": ");
+           SERIAL_PROTOCOL(bedlevelprobes[i]);
+           SERIAL_EOL;
+         }
+       }
+       
+       sample_avg = sample_sum >> 3;
+       if (verbose_level > 3) {
+         SERIAL_PROTOCOLPGM("sample_sum: ");
+         SERIAL_PROTOCOL(sample_sum);
+         SERIAL_EOL;
+         SERIAL_PROTOCOLPGM("sample_avg: ");
+         SERIAL_PROTOCOL(sample_avg);
+         SERIAL_EOL;
+       }
+       sample_sum = sample_avg + bedlevelprobes[8];
+       sample_avg = sample_sum >> 1;
+       if (verbose_level > 3) {
+         SERIAL_PROTOCOLPGM("sample_sum: ");
+         SERIAL_PROTOCOL(sample_sum);
+         SERIAL_EOL;
+         SERIAL_PROTOCOLPGM("sample_avg: ");
+         SERIAL_PROTOCOL(sample_avg);
+         SERIAL_EOL;
+       }
+       
+       if (verbose_level > 3) {
+         SERIAL_PROTOCOLPGM("Bed");
+         SERIAL_PROTOCOLPGM(" X: ");
+         SERIAL_PROTOCOL_F(x, 3);
+         SERIAL_PROTOCOLPGM(" Y: ");
+         SERIAL_PROTOCOL_F(y, 3);
```

Comment:
yeah I had some interesting experiences with `SERIAL_PROTOCOL_F`. Basically, it's supposed to take in a `float` and print it. If you give a float value, the 2nd arg tells it how many digits to print. If it's not a float, it represets the base of the number.